modules/root/protected_root_api.rb

```diff
@@ -0,0 +1,42 @@
+class Proxy::ProtectedRootApi < Sinatra::Base
+  helpers ::Proxy::Helpers
+
+  # authorize_with_trusted_hosts
+  # authorize_with_ssl_client
+
+  get "/v2/features" do
+    begin
+      enabled_plugins = ::Proxy::Plugins.instance.select do |p|
+        p[:name] != :foreman_proxy && \
+          p[:state] == :running && \
+          p[:class].ancestors.include?(::Proxy::Plugin)
+      end
+      content_type :json
+
+      attributes = %i[http_enabled https_enabled settings]
+
+      plugins = enabled_plugins.inject({}) do |hash, plugin|
+        result = plugin.slice(*attributes)
+        result[:capabilities] = process_capabilities(plugin[:capabilities])
+        hash[plugin[:name]] = result
+        hash
+      end
+
+      plugins.to_json
+    rescue => e
+      log_halt 400, e
+    end
+  end
+
+  def process_capabilities(capabilities)
+    capabilities = capabilities.map do |capability|
+      if capability.is_a?(Proc)
+        capability.call
```

Comment:
I like the ability of using a proc. I didn't think of that. Should we take some safe guards here with regards to exceptions or timeouts (if they call to external services)?

A possible test case in the current code would be to use puppet. If it has a provider set, then it should have the capability puppetrun. This can then be used by the foreman to detect it at runtime. Currently it's a setting in foreman's settings.yaml to enable the button globally but it's really a capability on the proxy for the puppet feature.

Reply:
I'd be OK with mentioning in docs that you are expected to handle all exceptions and timeouts.